File: src/challenge5.cpp
#include <fstream>
#include <string>
#include <vector>

#include "ByteSequence.hpp"
#include "CharacterFrequencyScoreCalculator.hpp"
#include "Encryptor.hpp"

int main(int argc, char *argv[]) {
  std::string text = "Burning 'em, if you ain't quick and nimble\nI go crazy "
                     "when I hear a cymbal";
  ByteSequence textByteSequence;
  textByteSequence.initializeFromAsciiBytes(
      std::vector<char>(&text[0], &text[text.size()]));

  std::string key = "ICE";
  ByteSequence keyByteSequence;
  keyByteSequence.initializeFromAsciiBytes(
      std::vector<char>(&key[0], &key[key.size()]));

  auto encryptedByteSequence =
      Encryptor::getRepeatingKeyXorEncryptedByteSequence(textByteSequence,
                                                         keyByteSequence);

  encryptedByteSequence.printHexEncodedAsciiString();

  return 0;
}

File: challenge_5.rb
require 'pry-byebug'

require_relative 'challenge_1.rb'
require_relative 'challenge_2.rb'

module Crypt
  def self.repeating_key_xor(plaintext, key)
    plaintext.bytes.each_slice(key.size).map do |slice|
      xor(slice.pack('C*'), key)
    end.join('')
  end
end

plaintext = "Burning 'em, if you ain't quick and nimble\nI go crazy when I hear a cymbal"
p encode_hex(Crypt.repeating_key_xor(plaintext, 'ICE'))
